File: Homework/Hoang_Nguyen/lesson7/687_longestUnivaluePath.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestUnivaluePath = function (root) {
  let maxPath = 0;
  dfs(root);
  return maxPath;

  function dfs(node) {
    if (!node) return 0;

    const left = dfs(node.left);
    const right = dfs(node.right);
    let maxLeft = 0, // max path from left node
      maxRight = 0; // max path from right node
    if (node.left && node.left.val === node.val) {
      maxLeft = left + 1;
    }
    if (node.right && node.right.val === node.val) {
      maxRight = right + 1;
    }
    maxPath = Math.max(maxPath, maxLeft + maxRight);
    return Math.max(maxLeft, maxRight);
  }
};

File: Homework/Ngan_Luu/687-longest-univalue-path.py
# 687. https://leetcode.com/problems/longest-univalue-path/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def longestUnivaluePath(self, root: Optional[TreeNode]) -> int:
        def dfs(node,prev):
            if not(node):
                return 0
            left=dfs(node.left,node.val)
            right=dfs(node.right,node.val)
            longestPath[0]=max(longestPath[0],left+right)
            if node.val==prev:
                return max(left,right)+1
            else:
                return 0
        longestPath=[0]
        if root:
            dfs(root,root.val)
            return longestPath[0]
        else:
            return 0